flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSource.java

```diff
@@ -107,7 +106,11 @@ public void run(final Object lockingObject,
 				// in theory, the subclasses of StreamSource may implement the BoundedOneInput interface,
 				// so we still need the following call to end the input
 				synchronized (lockingObject) {
-					operatorChain.endMainOperatorInput(1);
+					if (this instanceof BoundedOneInput) {
+						((BoundedOneInput) this).endInput();
+					} else if (this instanceof BoundedMultiInput) {
+						((BoundedMultiInput) this).endInput(1);
```

Comment:
This looks tricky. The source might run at other positions than input 1.

Reply:
Not sure if I understand the comment.

Does a StreamSource even have inputs? Of course I might be missing something here, but isn't the input id, the id of an input that has finished? Honestly, I find this part of code quite shady, but I think the logic pre and post this change is the same. Again I might be wrong.

Reply:
Ah, maybe I confused this (and I also find that part of the code not very easy).

Is this the input of the source operator (there should be none)? I understood this was the input of the successor to the source operator, meaning the source tells its downstream successor that the input that it pushes to is now done.